src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs

```diff
@@ -0,0 +1,344 @@
+
+            message = (logger, arg1, arg2, arg3, exception) =>
+            {
+                if (logger.IsEnabled(logLevel))
+                {
+                    logger.Log(logLevel, eventId, new LogValues<string, T1, T2, T3>(formatter, eventName, arg1, arg2, arg3), exception, LogValues<string, T1, T2, T3>.Callback);
+                }
+            };
+        }
+
+        private class LogValues : ILogValues
+        {
+            public static Func<object, Exception, string> Callback = (state, exception) => ((LogValues)state)._formatter.Format(((LogValues)state).ToArray());
+
+            private static IEnumerable<KeyValuePair<string, object>> _getValues = new KeyValuePair<string, object>[0];
+            private static object[] _toArray = new object[0];
+
+            private readonly LogValuesFormatter _formatter;
+
+            public LogValues(LogValuesFormatter formatter)
+            {
+                _formatter = formatter;
+            }
+
+            public IEnumerable<KeyValuePair<string, object>> GetValues() => _getValues;
+
+            public object[] ToArray() => _toArray;
+
+            public override string ToString() => _formatter.Format(ToArray());
+        }
+
+        private class LogValues<T0> : ILogValues
+        {
+            public static Func<object, Exception, string> Callback = (state, exception) => ((LogValues<T0>)state)._formatter.Format(((LogValues<T0>)state).ToArray());
+
+            private readonly LogValuesFormatter _formatter;
+            private readonly T0 _value0;
+
+            public LogValues(LogValuesFormatter formatter, T0 value0)
+            {
+                _formatter = formatter;
+                _value0 = value0;
+            }
+
+            public IEnumerable<KeyValuePair<string, object>> GetValues() => new[]
+            {
+                new KeyValuePair<string, object>(_formatter.ValueNames[0], _value0),
+                new KeyValuePair<string, object>("OriginalFormat", _formatter.OriginalFormat),
+            };
+
+            public object[] ToArray() => new object[] { _value0 };
+
+            public override string ToString() => _formatter.Format(ToArray());
+        }
+
+        private class LogValues<T0, T1> : ILogValues
+        {
+            public static Func<object, Exception, string> Callback = (state, exception) => ((LogValues<T0, T1>)state)._formatter.Format(((LogValues<T0, T1>)state).ToArray());
+
+            private readonly LogValuesFormatter _formatter;
+            private readonly T0 _value0;
+            private readonly T1 _value1;
+
+            public LogValues(LogValuesFormatter formatter, T0 value0, T1 value1)
+            {
+                _formatter = formatter;
+                _value0 = value0;
+                _value1 = value1;
+            }
+
+            public IEnumerable<KeyValuePair<string, object>> GetValues() => new[]
```

Comment:
@loudej @lodejard wondering why do we need to do it this way...like we could use the formatter which already does this for us `formatter.GetValues(ToArray())`
https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs#L146